src/persist-client/src/internal/encoding.rs

```diff
@@ -905,13 +905,25 @@ impl<T: Timestamp + Codec64> RustType<ProtoHollowBatch> for HollowBatch<T> {
                 .map(|key| HollowBatchPart {
                     key: PartialBatchKey(key),
                     encoded_size_bytes: 0,
+                    len: 0,
                     stats: None,
                 }),
         );
+
+        // We didn't always fill in len for parts, so historical data will give
+        // back the proto default of 0. We don't write parts with 0 updates, so
+        // conveniently, we can use this as a sentinel for old data. Fill in the
```

Comment:
we are guaranteed to never have a part with 0 updates, right?

Reply:
yeah, this line skips adding it to the batch if it's empty: https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/batch.rs#L467-L469. this also happens to be a pretty critical performance optimization where we don't touch s3 for an empty compare_and_append